sn_client/src/wallet.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2023 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::Client;
+
+use sn_transfers::{
+    client_transfers::TransactionOutputs,
+    wallet::{Error, LocalWallet, Result},
+};
+
+use futures::future::join_all;
+use sn_dbc::{Dbc, PublicAddress, Token};
+
+/// A wallet client can be used to send and
+/// receive tokens to/from other wallets.
+pub struct WalletClient {
+    client: Client,
+    wallet: LocalWallet,
+    /// These have not yet been successfully confirmed in
+    /// the network and need to be republished, to reach network validity.
+    /// We maintain the order they were added in, as to republish
+    /// them in the correct order, in case any later spend was
+    /// dependent on an earlier spend.
+    unconfirmed_txs: Vec<TransactionOutputs>,
+}
+
+impl WalletClient {
+    /// Create a new wallet client.
+    pub fn new(client: Client, wallet: LocalWallet) -> Self {
+        Self {
+            client,
+            wallet,
+            unconfirmed_txs: vec![],
+        }
+    }
+
+    /// Send tokens to another wallet.
+    pub async fn send(&mut self, amount: Token, to: PublicAddress) -> Result<Dbc> {
+        // retry previous failures
+        self.resend_pending_txs().await;
+
+        // offline transfer
+        let transfer = self.wallet.local_send(vec![(amount, to)]).await?;
+        let dbcs = transfer.created_dbcs.clone();
+
+        // send to network
+        println!("Sending transfer to the network: {transfer:#?}");
+        if let Err(error) = self.client.send(transfer.clone()).await {
+            println!("The transfer was not successfully registered in the network: {error:?}. It will be retried later.");
+            self.unconfirmed_txs.push(transfer);
+        }
+
+        // return created DBCs even if network part failed???
```

Comment:
Hi Anselme!

As you say, this is not totally clear.

Above we don't yet have an evaluation of the error returned. It could be transient errors, or something that means the transfer can never be registered in the network for some reason (something invalid, or double spend, but the actual real paths I haven't looked at now).

All errors there lead to the transfer being added to pending list.
There are two options there, or maybe three, and they are mostly about UX.

`A.` Ensure it's only adding to pending list on transient/recoverable errors.
`B.` Allow removing transfers from the pending list (if by inspection it can be determined that it can't be registered, or... also even if user doesn't want to send it anymore).
`C.` Combine the two above.

So. Let's say it's a transient error.
The user still wants the created dbcs (and store in wallet, so they are not lost). User can decide if they want to resend the pending tx and confirm it before handing over the created dbcs. Or user hands over the dbcs and tries confirming later. (Depends on what the recipient/sender agrees on of course).

If it's not a transient error, then I think it's appropriate to handle that above at `ln 54`, and not add the tx to the pending list.
(However the UX might be improved by keeping the failed tx somewhere anyway, just as a history of actions - together with the error description/reason why it failed).
In this case, I also think the created dbcs should be dropped without ever being exposed to user, as they will never become valid.